File: src/semantic-validation.js
import { SemanticExtractionSchema } from "./schemas.js";

export const SEMANTIC_VALIDATION_CODES = Object.freeze({
  SCHEMA_INVALID: "schema_invalid",
  EVIDENCE_OUT_OF_BOUNDS: "evidence_out_of_bounds",
  EVIDENCE_TEXT_MISMATCH: "evidence_text_mismatch",
  DURABILITY_REJECTED: "durability_rejected",
  USER_FACING_FIRST_PERSON: "user_facing_first_person",
  USER_FACING_USER_MISSING: "user_facing_user_missing",
  USER_FACING_IMPLEMENTATION_DATE: "user_facing_implementation_date",
  OCCURRED_AT_TEXT_MISSING: "occurred_at_text_missing",
  OCCURRED_AT_TEXT_MISMATCH: "occurred_at_text_mismatch",
  EVIDENCE_INDEX_OUT_OF_BOUNDS: "evidence_index_out_of_bounds",
  ENTITY_MENTION_UNSUPPORTED: "entity_mention_unsupported",
  DUPLICATE_ENTITY: "duplicate_entity",
  RELATIONSHIP_ENDPOINT_UNSUPPORTED: "relationship_endpoint_unsupported",
  DUPLICATE_RELATIONSHIP: "duplicate_relationship",
  DUPLICATE_ATOM: "duplicate_atom",
  BOUNDARY_MULTIPLE_SENTENCES: "boundary_multiple_sentences",
  BOUNDARY_MULTIPLE_SUBJECTS: "boundary_multiple_subjects",
  BOUNDARY_BROAD_SOURCE_COVERAGE: "boundary_broad_source_coverage",
  BOUNDARY_INDEPENDENT_AND_CLAIMS: "boundary_independent_and_claims",
});

export const SEMANTIC_ACCEPTANCE_THRESHOLDS = Object.freeze({
  durability: 0.70,
  entity: 0.70,
  relationship: 0.75,
  secondaryType: 0.15,
});

const EMPTY_DROP_COUNTS = Object.freeze({
  entities: 0,
  relationships: 0,
  types: 0,
  actions: 0,
  topics: 0,
});

// Keep uppercase acronyms and Roman numerals such as "US" and "World War I"
// from being mistaken for first-person pronouns.
const FIRST_PERSON_PATTERN =
  /\b(?:I|[Mm]e|[Mm]y|[Mm]ine|[Mm]yself|[Ww]e|[Uu]s|[Oo]ur|[Oo]urs|[Oo]urselves|[Ss]elf|[Tt]he speaker)\b/;
const SOURCE_AUTHOR_PATTERN =
  /\b(?:I|[Mm]e|[Mm]y|[Mm]ine|[Mm]yself|[Ww]e|[Uu]s|[Oo]ur|[Oo]urs|[Oo]urselves)\b/;

export class SemanticValidationError extends Error {
  constructor(result) {
    const details = result.issues
      .map(({ code, path, message }) =>
        `${code}${path.length ? ` at ${path.join(".")}` : ""}: ${message}`)
      .join("; ");
    super(`Semantic extraction failed deterministic validation: ${details}`);
    this.name = "SemanticValidationError";
    this.code = "SEMANTIC_EXTRACTION_INVALID";
    this.issues = result.issues;
    this.warnings = result.warnings;
    this.dropCounts = result.dropCounts;
    this.extraction = result.extraction;
  }
}

export function validateSemanticExtraction(sourceText, candidate) {
  if (typeof sourceText !== "string") {
    throw new TypeError("sourceText must be a string");
  }

  const parsed = SemanticExtractionSchema.safeParse(candidate);
  if (!parsed.success) {
    return {
      success: false,
      extraction: null,
      issues: parsed.error.issues.map((issue) => ({
        code: SEMANTIC_VALIDATION_CODES.SCHEMA_INVALID,
        path: issue.path,
        message: issue.message,
      })),
      warnings: [],
      dropCounts: { ...EMPTY_DROP_COUNTS },
    };
  }

  const warnings = [];
  const rejectedIssues = [];
  const dropCounts = { ...EMPTY_DROP_COUNTS };
  const accepted = [];

  for (const [atomIndex, memory] of parsed.data.memories.entries()) {
    const atomPath = ["memories", atomIndex];
    const normalized = normalizeMemory(
      sourceText,
      memory,
      atomPath,
      warnings,
      dropCounts,
    );

    if (
      !normalized.durability.durable
      || normalized.durability.confidence
        < SEMANTIC_ACCEPTANCE_THRESHOLDS.durability
    ) {
      warnings.push(issue(
        SEMANTIC_VALIDATION_CODES.DURABILITY_REJECTED,
        [...atomPath, "durability"],
        "dropped atom below the durable-memory acceptance threshold",
      ));
      continue;
    }

    const coreIssues = validateCoreMemory(sourceText, normalized, atomPath);
    if (coreIssues.length > 0) {
      rejectedIssues.push(...coreIssues);
      continue;
    }
    accepted.push(normalized);
  }

  const memories = deduplicateMemories(accepted, warnings);
  const noUsableMemory = parsed.data.memories.length > 0
    && memories.length === 0
    && rejectedIssues.length > 0;

  return {
    success: !noUsableMemory,
    extraction: { memories },
    issues: noUsableMemory ? rejectedIssues : [],
    warnings: [...warnings, ...(noUsableMemory ? [] : rejectedIssues)],
    dropCounts,
  };
}

export function assertValidSemanticExtraction(sourceText, candidate, options) {
  const result = validateSemanticExtraction(sourceText, candidate, options);
  if (!result.success) throw new SemanticValidationError(result);
  return result;
}

function normalizeMemory(sourceText, memory, atomPath, warnings, dropCounts) {
  const types = normalizeTypes(memory.types, warnings, dropCounts, atomPath);
  const entities = normalizeEntities(memory, warnings, dropCounts, atomPath);
  const relationships = normalizeRelationships(
    memory,
    entities,
    warnings,
    dropCounts,
    atomPath,
  );
  const actions = normalizeUniqueStrings(memory.actions, "actions", dropCounts);
  const topics = normalizeUniqueStrings(memory.topics, "topics", dropCounts);
  const occurredAt = normalizeOccurredAt(sourceText, memory.occurredAt, warnings, atomPath);
  const summary = FIRST_PERSON_PATTERN.test(memory.summary)
    ? memory.text
    : memory.summary;

  if (summary !== memory.summary) {
    warnings.push(issue(
      SEMANTIC_VALIDATION_CODES.USER_FACING_FIRST_PERSON,
      [...atomPath, "summary"],
      "replaced non-third-person summary with memory text",
    ));
  }

  return {
    ...memory,
    summary,
    types,
    occurredAt,
    entities,
    relationships,
    actions,
    topics,
  };
}

function normalizeTypes(types, warnings, dropCounts, atomPath) {
  const dominantTypeIndex = types.reduce(
    (dominant, type, index) =>
      dominant === -1 || type.weight > types[dominant].weight ? index : dominant,
    -1,
  );
  return types.filter((type, index) => {
    const accepted = index === dominantTypeIndex
      || type.weight >= SEMANTIC_ACCEPTANCE_THRESHOLDS.secondaryType;
    if (!accepted) {
      dropCounts.types += 1;
      warnings.push(issue(
        SEMANTIC_VALIDATION_CODES.SCHEMA_INVALID,
        [...atomPath, "types", index],
        "dropped weak secondary memory type",
      ));
    }
    return accepted;
  });
}

function normalizeEntities(memory, warnings, dropCounts, atomPath) {
  const entities = [];
  const seen = new Set();

  for (const [entityIndex, entity] of memory.entities.entries()) {
    const path = [...atomPath, "entities", entityIndex];
    if (entity.confidence < SEMANTIC_ACCEPTANCE_THRESHOLDS.entity) {
      dropCounts.entities += 1;
      continue;
    }

    const evidence = resolveEvidence(memory, entity.evidenceSpanIndexes);
    const mention = evidence && findLiteral(evidence.map(({ text }) => text), entity.mention);
    if (!mention) {
      dropCounts.entities += 1;
      warnings.push(issue(
        SEMANTIC_VALIDATION_CODES.ENTITY_MENTION_UNSUPPORTED,
        [...path, "mention"],
        "dropped entity whose mention was not present in cited evidence",
      ));
      continue;
    }

    const normalized = { ...entity, mention };
    const key = [normalized.mention, normalized.kind, normalized.canonicalName || ""]
      .map(normalizeKey)
      .join("\u0000");
    if (seen.has(key)) {
      dropCounts.entities += 1;
      warnings.push(issue(
        SEMANTIC_VALIDATION_CODES.DUPLICATE_ENTITY,
        path,
        "dropped duplicate entity",
      ));
      continue;
    }
    seen.add(key);
    entities.push(normalized);
  }

  return entities;
}

function normalizeRelationships(memory, entities, warnings, dropCounts, atomPath) {
  const entityNames = new Map();
  for (const entity of entities) {
    for (const name of [entity.mention, entity.canonicalName]) {
      if (name) entityNames.set(normalizeKey(name), name);
    }
  }

  const relationships = [];
  const seen = new Set();
  for (const [relationshipIndex, relationship] of memory.relationships.entries()) {
    const path = [...atomPath, "relationships", relationshipIndex];
    if (relationship.confidence < SEMANTIC_ACCEPTANCE_THRESHOLDS.relationship) {
      dropCounts.relationships += 1;
      continue;
    }

    const evidence = resolveEvidence(memory, relationship.evidenceSpanIndexes);
    if (!evidence) {
      dropCounts.relationships += 1;
      warnings.push(issue(
        SEMANTIC_VALIDATION_CODES.EVIDENCE_INDEX_OUT_OF_BOUNDS,
        [...path, "evidenceSpanIndexes"],
        "dropped relationship with invalid evidence indexes",
      ));
      continue;
    }

    const subject = normalizeRelationshipEndpoint(
      relationship.subject,
      entityNames,
      evidence,
    );
    const object = normalizeRelationshipEndpoint(
      relationship.object,
      entityNames,
      evidence,
    );
    if (!subject || !object) {
      dropCounts.relationships += 1;
      warnings.push(issue(
        SEMANTIC_VALIDATION_CODES.RELATIONSHIP_ENDPOINT_UNSUPPORTED,
        path,
        "dropped relationship with an unsupported endpoint",
      ));
      continue;
    }

    const normalized = { ...relationship, subject, object };
    const key = [subject, normalized.predicate, object]
      .map(normalizeKey)
      .join("\u0000");
    if (seen.has(key)) {
      dropCounts.relationships += 1;
      warnings.push(issue(
        SEMANTIC_VALIDATION_CODES.DUPLICATE_RELATIONSHIP,
        path,
        "dropped duplicate relationship",
      ));
      continue;
    }
    seen.add(key);
    relationships.push(normalized);
  }

  return relationships;
}

function normalizeRelationshipEndpoint(value, entityNames, evidence) {
  const key = normalizeKey(value);
  if (["self", "user", "the user", "user's"].includes(key)) {
    return "self";
  }
  if (entityNames.has(key)) return entityNames.get(key);
  return findLiteral(evidence.map(({ text }) => text), value);
}

function normalizeOccurredAt(sourceText, occurredAt, warnings, atomPath) {
  if (!occurredAt.text) {
    if (occurredAt.normalized === null) return occurredAt;
    warnings.push(issue(
      SEMANTIC_VALIDATION_CODES.OCCURRED_AT_TEXT_MISSING,
      [...atomPath, "occurredAt"],
      "cleared occurrence date without source text evidence",
    ));
    return { text: "", normalized: null, confidence: 0 };
  }

  const text = findLiteral([sourceText], occurredAt.text);
  if (text) return { ...occurredAt, text };
  warnings.push(issue(
    SEMANTIC_VALIDATION_CODES.OCCURRED_AT_TEXT_MISMATCH,
    [...atomPath, "occurredAt"],
    "cleared occurrence date not found in source text",
  ));
  return { text: "", normalized: null, confidence: 0 };
}

function validateCoreMemory(sourceText, memory, atomPath) {
  const issues = [];
  for (const [spanIndex, span] of memory.evidenceSpans.entries()) {
    const path = [...atomPath, "evidenceSpans", spanIndex];
    if (span.end > sourceText.length) {
      issues.push(issue(
        SEMANTIC_VALIDATION_CODES.EVIDENCE_OUT_OF_BOUNDS,
        path,
        `span end ${span.end} exceeds source length ${sourceText.length}`,
      ));
    }
    if (sourceText.slice(span.start, span.end) !== span.text) {
      issues.push(issue(
        SEMANTIC_VALIDATION_CODES.EVIDENCE_TEXT_MISMATCH,
        [...path, "text"],
        "span text must exactly equal the source slice",
      ));
    }
  }

  if (FIRST_PERSON_PATTERN.test(memory.text)) {
    issues.push(issue(
      SEMANTIC_VALIDATION_CODES.USER_FACING_FIRST_PERSON,
      [...atomPath, "text"],
      "text must refer to the memory owner in third person as user or user's",
    ));
  }
  const hasFirstPersonEvidence = memory.evidenceSpans.some(({ text }) =>
    SOURCE_AUTHOR_PATTERN.test(text)
  );
  if (hasFirstPersonEvidence && !/\buser\b/i.test(memory.text)) {
    issues.push(issue(
      SEMANTIC_VALIDATION_CODES.USER_FACING_USER_MISSING,
      [...atomPath, "text"],
      "text must identify the first-person source author as user",
    ));
  }
  return issues;
}

function resolveEvidence(memory, indexes) {
  if (indexes.some((index) => index >= memory.evidenceSpans.length)) return null;
  return indexes.map((index) => memory.evidenceSpans[index]);
}

function findLiteral(texts, value) {
  const needle = String(value).trim();
  if (!needle) return null;
  for (const text of texts) {
    const exactIndex = text.indexOf(needle);
    if (exactIndex >= 0) return text.slice(exactIndex, exactIndex + needle.length);
    const foldedIndex = text.toLocaleLowerCase().indexOf(needle.toLocaleLowerCase());
    if (foldedIndex >= 0) return text.slice(foldedIndex, foldedIndex + needle.length);
  }
  return null;
}

function normalizeUniqueStrings(values, field, dropCounts) {
  const seen = new Set();
  const result = [];
  for (const value of values) {
    const trimmed = value.trim();
    const key = normalizeKey(trimmed);
    if (seen.has(key)) {
      dropCounts[field] += 1;
      continue;
    }
    seen.add(key);
    result.push(trimmed);
  }
  return result;
}

function deduplicateMemories(memories, warnings) {
  const seen = new Set();
  return memories.filter((memory, index) => {
    const key = normalizeKey(memory.text);
    if (!seen.has(key)) {
      seen.add(key);
      return true;
    }
    warnings.push(issue(
      SEMANTIC_VALIDATION_CODES.DUPLICATE_ATOM,
      ["memories", index, "text"],
      "dropped duplicate atom",
    ));
    return false;
  });
}

function normalizeKey(value) {
  return String(value).normalize("NFKC").trim().replace(/\s+/g, " ").toLocaleLowerCase();
}

function issue(code, path, message) {
  return { code, path, message };
}

File: atlas-py/atlas_py/semantic_validation.py
"""Deterministic validation of V3 semantic extraction output.

Port of ``src/semantic-validation.js``. The validator is intentionally
strict: it inspects evidence spans against the source text, drops
low-confidence entities/relationships, and surfaces boundary issues
that indicate the LLM split the source text incorrectly.
"""

from __future__ import annotations

import re
import unicodedata
from types import MappingProxyType
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, ValidationError

from .schemas import SemanticExtractionSchema


SEMANTIC_VALIDATION_CODES: MappingProxyType = MappingProxyType({
    "SCHEMA_INVALID": "schema_invalid",
    "EVIDENCE_OUT_OF_BOUNDS": "evidence_out_of_bounds",
    "EVIDENCE_TEXT_MISMATCH": "evidence_text_mismatch",
    "DURABILITY_REJECTED": "durability_rejected",
    "USER_FACING_INTERNAL_SUBJECT": "user_facing_internal_subject",
    "USER_FACING_IMPLEMENTATION_DATE": "user_facing_implementation_date",
    "OCCURRED_AT_TEXT_MISSING": "occurred_at_text_missing",
    "OCCURRED_AT_TEXT_MISMATCH": "occurred_at_text_mismatch",
    "EVIDENCE_INDEX_OUT_OF_BOUNDS": "evidence_index_out_of_bounds",
    "ENTITY_MENTION_UNSUPPORTED": "entity_mention_unsupported",
    "DUPLICATE_ENTITY": "duplicate_entity",
    "RELATIONSHIP_ENDPOINT_UNSUPPORTED": "relationship_endpoint_unsupported",
    "DUPLICATE_RELATIONSHIP": "duplicate_relationship",
    "DUPLICATE_ATOM": "duplicate_atom",
    "BOUNDARY_MULTIPLE_SENTENCES": "boundary_multiple_sentences",
    "BOUNDARY_MULTIPLE_SUBJECTS": "boundary_multiple_subjects",
    "BOUNDARY_BROAD_SOURCE_COVERAGE": "boundary_broad_source_coverage",
    "BOUNDARY_INDEPENDENT_AND_CLAIMS": "boundary_independent_and_claims",
})

SEMANTIC_ACCEPTANCE_THRESHOLDS: MappingProxyType = MappingProxyType({
    "durability": 0.70,
    "entity": 0.70,
    "relationship": 0.75,
    "secondaryType": 0.15,
})

EMPTY_DROP_COUNTS: MappingProxyType = MappingProxyType({
    "entities": 0,
    "relationships": 0,
    "types": 0,
    "actions": 0,
    "topics": 0,
})


class SemanticValidationError(Exception):
    def __init__(self, result: dict):
        details = "; ".join(
            f"{issue['code']}{' at ' + '.'.join(str(p) for p in issue['path']) if issue['path'] else ''}: {issue['message']}"
            for issue in result["issues"]
        )
        super().__init__(f"Semantic extraction failed deterministic validation: {details}")
        self.code = "SEMANTIC_EXTRACTION_INVALID"
        self.issues = result["issues"]
        self.drop_counts = result["drop_counts"]
        self.extraction = result["extraction"]


def _normalize_key(value) -> str:
    return (
        unicodedata.normalize("NFKC", str(value))
        .strip()
        .replace("\u00a0", " ")
    ).lower()


def _normalize_unique_strings(values, field: str, drop_counts: dict) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for value in values or []:
        text = str(value).strip()
        key = _normalize_key(text)
        if key in seen:
            drop_counts[field] += 1
            continue
        seen.add(key)
        out.append(text)
    return out


def _add_issue(issues: list, code: str, path: list, message: str) -> None:
    issues.append({"code": code, "path": list(path), "message": message})


def _apply_acceptance_policy(memory: dict, drop_counts: dict) -> dict:
    all_entity_names = {
        _normalize_key(item)
        for entity in memory.get("entities", [])
        for item in (entity.get("mention"), entity.get("canonicalName"))
        if item
    }
    accepted_entities = []
    for entity in memory.get("entities", []):
        if entity.get("confidence", 0) >= SEMANTIC_ACCEPTANCE_THRESHOLDS["entity"]:
            accepted_entities.append(entity)
        else:
            drop_counts["entities"] += 1
    entity_names = {
        _normalize_key(name)
        for entity in accepted_entities
        for name in (entity.get("mention"), entity.get("canonicalName"))
        if name
    }
    accepted_relationships = []
    for relationship in memory.get("relationships", []):
        subject_dropped = _endpoint_references_dropped_entity(
            relationship.get("subject"), all_entity_names, entity_names
        )
        object_dropped = _endpoint_references_dropped_entity(
            relationship.get("object"), all_entity_names, entity_names
        )
        if subject_dropped or object_dropped:
            drop_counts["relationships"] += 1
            continue
        if relationship.get("confidence", 0) < SEMANTIC_ACCEPTANCE_THRESHOLDS["relationship"]:
            drop_counts["relationships"] += 1
            continue
        accepted_relationships.append(relationship)

    # Keep the highest-weight type plus any secondary type above the threshold.
    types = memory.get("types", [])
    dominant_index = -1
    for index, type_entry in enumerate(types):
        if dominant_index == -1 or type_entry.get("weight", 0) > types[dominant_index].get("weight", 0):
            dominant_index = index
    accepted_types = []
    for index, type_entry in enumerate(types):
        if index == dominant_index or type_entry.get("weight", 0) >= SEMANTIC_ACCEPTANCE_THRESHOLDS["secondaryType"]:
            accepted_types.append(type_entry)
        else:
            drop_counts["types"] += 1

    actions = _normalize_unique_strings(memory.get("actions", []), "actions", drop_counts)
    topics = _normalize_unique_strings(memory.get("topics", []), "topics", drop_counts)

    return {
        **memory,
        "entities": accepted_entities,
        "relationships": accepted_relationships,
        "types": accepted_types,
        "actions": actions,
        "topics": topics,
    }


def _endpoint_references_dropped_entity(endpoint, all_entity_names: set, accepted_entity_names: set) -> bool:
    if endpoint is None:
        return False
    key = _normalize_key(endpoint)
    return key in all_entity_names and key not in accepted_entity_names


def _validate_atom(source_text: str, memory: dict, atom_index: int, issues: list, boundary_audit: bool) -> None:
    atom_path = ["memories", atom_index]
    spans = memory.get("evidenceSpans", [])
    for span_index, span in enumerate(spans):
        path = [*atom_path, "evidenceSpans", span_index]
        if span.get("end", 0) > len(source_text):
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["EVIDENCE_OUT_OF_BOUNDS"],
                path,
                f"span end {span['end']} exceeds source length {len(source_text)}",
            )
        if source_text[span.get("start", 0):span.get("end", 0)] != span.get("text"):
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["EVIDENCE_TEXT_MISMATCH"],
                [*path, "text"],
                "span text must exactly equal the source slice",
            )

    durability = memory.get("durability", {})
    if (
        not durability.get("durable")
        or durability.get("confidence", 0) < SEMANTIC_ACCEPTANCE_THRESHOLDS["durability"]
    ):
        _add_issue(
            issues,
            SEMANTIC_VALIDATION_CODES["DURABILITY_REJECTED"],
            [*atom_path, "durability"],
            "atom does not meet the durable-memory acceptance threshold",
        )

    _validate_user_facing_prose(memory, atom_path, issues)
    _validate_occurred_at(source_text, memory.get("occurredAt", {}), atom_path, issues)
    _validate_entities(memory, atom_path, issues)
    _validate_relationships(memory, atom_path, issues)
    if boundary_audit:
        _audit_boundaries(source_text, memory, atom_path, issues)


_USER_FACING_INTERNAL = re.compile(r"\b(?:self|the speaker|the user)\b", re.IGNORECASE)
_USER_FACING_DATE = re.compile(r"\b(?:the )?(?:ingestion|source|current|today'?s) date\b", re.IGNORECASE)


def _validate_user_facing_prose(memory: dict, atom_path: list, issues: list) -> None:
    for field in ("text", "summary"):
        value = memory.get(field, "") or ""
        if _USER_FACING_INTERNAL.search(value):
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["USER_FACING_INTERNAL_SUBJECT"],
                [*atom_path, field],
                f"{field} must use first-person language; self is reserved for relationships",
            )
        if _USER_FACING_DATE.search(value):
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["USER_FACING_IMPLEMENTATION_DATE"],
                [*atom_path, field],
                f"{field} must use the source time phrase or resolved calendar date",
            )


def _validate_occurred_at(source_text: str, occurred_at: dict, atom_path: list, issues: list) -> None:
    text = occurred_at.get("text", "") or ""
    if not text:
        if occurred_at.get("normalized") is not None:
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["OCCURRED_AT_TEXT_MISSING"],
                [*atom_path, "occurredAt", "text"],
                "a normalized occurrence date requires an exact source time phrase",
            )
        return
    if text not in source_text:
        _add_issue(
            issues,
            SEMANTIC_VALIDATION_CODES["OCCURRED_AT_TEXT_MISMATCH"],
            [*atom_path, "occurredAt", "text"],
            "occurredAt.text must be an exact source substring",
        )


def _resolve_evidence(memory: dict, indexes, path: list, issues: list) -> list[dict]:
    evidence: list[dict] = []
    for position, index in enumerate(indexes or []):
        if index >= len(memory.get("evidenceSpans", [])):
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["EVIDENCE_INDEX_OUT_OF_BOUNDS"],
                [*path, "evidenceSpanIndexes", position],
                f"evidence span index {index} does not exist",
            )
            continue
        evidence.append(memory["evidenceSpans"][index])
    return evidence


def _validate_entities(memory: dict, atom_path: list, issues: list) -> None:
    seen: set[str] = set()
    for entity_index, entity in enumerate(memory.get("entities", [])):
        path = [*atom_path, "entities", entity_index]
        evidence = _resolve_evidence(
            memory, entity.get("evidenceSpanIndexes"), path, issues
        )
        if not any(entity.get("mention", "") in span.get("text", "") for span in evidence):
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["ENTITY_MENTION_UNSUPPORTED"],
                [*path, "mention"],
                "entity mention must occur exactly inside cited evidence",
            )
        key = "\u0000".join(
            _normalize_key(part)
            for part in [entity.get("mention", ""), entity.get("kind", ""), entity.get("canonicalName") or ""]
        )
        if key in seen:
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["DUPLICATE_ENTITY"],
                path,
                "duplicate entity after normalization",
            )
        seen.add(key)


def _validate_relationships(memory: dict, atom_path: list, issues: list) -> None:
    entity_names = {
        _normalize_key(name)
        for entity in memory.get("entities", [])
        for name in (entity.get("mention"), entity.get("canonicalName"))
        if name
    }
    seen: set[str] = set()
    for relationship_index, relationship in enumerate(memory.get("relationships", [])):
        path = [*atom_path, "relationships", relationship_index]
        evidence = _resolve_evidence(
            memory, relationship.get("evidenceSpanIndexes"), path, issues
        )
        for field in ("subject", "object"):
            endpoint = relationship.get(field, "") or ""
            if _normalize_key(endpoint) == "self":
                continue
            if _normalize_key(endpoint) in entity_names:
                continue
            if any(endpoint in span.get("text", "") for span in evidence):
                continue
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["RELATIONSHIP_ENDPOINT_UNSUPPORTED"],
                [*path, field],
                f"{field} must resolve to self, an extracted entity, or an exact evidence literal",
            )
        key = "\u0000".join(
            _normalize_key(part)
            for part in [relationship.get("subject", ""), relationship.get("predicate", ""), relationship.get("object", "")]
        )
        if key in seen:
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["DUPLICATE_RELATIONSHIP"],
                path,
                "duplicate relationship after normalization",
            )
        seen.add(key)


_SENTENCE_PATTERN = re.compile(r"[^.!?]+[.!?]+(?:\s+|$)")

_STATE_PREDICATES = {
    "lives_in", "works_at", "prefers", "related_to", "uses", "scheduled_for",
}


def _sentence_count(text: str) -> int:
    return len(_SENTENCE_PATTERN.findall(str(text)))


def _audit_boundaries(source_text: str, memory: dict, atom_path: list, issues: list) -> None:
    if _sentence_count(memory.get("text", "")) > 1:
        _add_issue(
            issues,
            SEMANTIC_VALIDATION_CODES["BOUNDARY_MULTIPLE_SENTENCES"],
            [*atom_path, "text"],
            "atom contains multiple complete sentences",
        )
    subjects = {_normalize_key(rel.get("subject", "")) for rel in memory.get("relationships", [])}
    if len(subjects) > 1:
        _add_issue(
            issues,
            SEMANTIC_VALIDATION_CODES["BOUNDARY_MULTIPLE_SUBJECTS"],
            [*atom_path, "relationships"],
            "atom contains relationship claims about multiple subjects",
        )
    if _sentence_count(source_text) > 1 and source_text:
        covered = sum(max(0, span.get("end", 0) - span.get("start", 0))
                      for span in memory.get("evidenceSpans", []))
        if covered / len(source_text) >= 0.8:
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["BOUNDARY_BROAD_SOURCE_COVERAGE"],
                [*atom_path, "evidenceSpans"],
                "one atom covers most of a multi-sentence source",
            )
    predicates = {
        _normalize_key(rel.get("predicate", "")).replace(" ", "_")
        for rel in memory.get("relationships", [])
    }
    state_predicates = predicates & _STATE_PREDICATES
    if re.search(r"\band\b", memory.get("text", ""), re.IGNORECASE) and len(state_predicates) > 1:
        _add_issue(
            issues,
            SEMANTIC_VALIDATION_CODES["BOUNDARY_INDEPENDENT_AND_CLAIMS"],
            [*atom_path, "text"],
            "atom joins multiple independently evolving state predicates with 'and'",
        )


def _validate_duplicate_atoms(memories: list[dict], issues: list) -> None:
    seen: dict[str, int] = {}
    for index, memory in enumerate(memories):
        key = _normalize_key(memory.get("text", ""))
        if key in seen:
            _add_issue(
                issues,
                SEMANTIC_VALIDATION_CODES["DUPLICATE_ATOM"],
                ["memories", index, "text"],
                f"duplicates atom {seen[key]}",
            )
        else:
            seen[key] = index


def validate_semantic_extraction(source_text, candidate, *, boundary_audit: bool = True) -> dict:
    if not isinstance(source_text, str):
        raise TypeError("sourceText must be a string")

    try:
        parsed = SemanticExtractionSchema.model_validate(candidate)
    except ValidationError as error:
        return {
            "success": False,
            "extraction": None,
            "issues": [
                {
                    "code": SEMANTIC_VALIDATION_CODES["SCHEMA_INVALID"],
                    "path": [str(item) for item in issue["path"]],
                    "message": issue["msg"],
                }
                for issue in error.errors()
            ],
            "drop_counts": {key: 0 for key in EMPTY_DROP_COUNTS},
        }

    issues: list = []
    drop_counts = {key: 0 for key in EMPTY_DROP_COUNTS}
    normalized_memories = []
    for atom_index, memory in enumerate(parsed.memories):
        normalized = _apply_acceptance_policy(memory.model_dump(), drop_counts)
        _validate_atom(source_text, normalized, atom_index, issues, boundary_audit)
        normalized_memories.append(normalized)
    _validate_duplicate_atoms(normalized_memories, issues)

    return {
        "success": len(issues) == 0,
        "extraction": {"memories": normalized_memories},
        "issues": issues,
        "drop_counts": drop_counts,
    }


def assert_valid_semantic_extraction(source_text, candidate, **options) -> dict:
    result = validate_semantic_extraction(source_text, candidate, **options)
    if not result["success"]:
        raise SemanticValidationError(result)
    return result
